File: bindings/cpp/examples/pipeline/screen_record.cpp
// screen_record.cpp - pipeline + device capability: screen + mic -> encode ->
// MP4.
//
// Status: real. GpuDevice::create() (adr/0007-gpu-device-factory.md) drives
// real Zero-Copy Screen capture, and EncodeSession::writeFrameFromDesktopCapture
// (adr/pipeline/0005-capture-encode-bridge-c-abi.md) polls + pushes each frame
// in one native call — no intermediate VideoFrame, Zero-Copy for Screen's
// GPU-backed frames. AutoVideoEncoder::open's VideoEncoderConfig::gpuDevice +
// inputFormat = PixelFormat::Bgra8 negotiate the GPU-input-capable path before
// the bridge is ever called (DXGI delivers BGRA8, not the NV12 default).
//
// On a machine whose encoder backend accepts a GPU-configured open but
// rejects the GPU-input path once frames actually start flowing (a real,
// pre-existing WMF/DX11 limitation — see the Rust gpu_write_frame_smoke.rs
// test and the C/Node.js/C#/Python screen_record siblings), this gracefully
// skips instead of crashing.
//
// Flow:
//   1. Create a GpuDevice; open display 0 at 30 fps with it.
//   2. Open the microphone at 48 kHz (drained, not muxed — same as
//      ScreenRecord.cs; camera_record.cpp's two-track remux is a separate
//      flow this file doesn't duplicate).
//   3. Open the auto H.264 encoder at the negotiated geometry, sharing the
//      same GpuDevice.
//   4. Record ~3 s: bridge screen frames straight into the encode session,
//      drain mic.
//   5. finish() -> MP4 bytes written to screen_out.mp4 by the caller.

#include <mediaway/mediaway.hpp>

#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <optional>
#include <string>
#include <thread>

namespace {

// NO_BACKEND / UNSUPPORTED / NO_DEVICE are expected outcomes in the ABI, not
// hard failures: the caller degrades gracefully (screen only, or skip).
bool isExpectedUnavailable(const mediaway::Error& error) {
  return error.status() == mediaway::Status::NoBackend ||
         error.status() == mediaway::Status::Unsupported ||
         error.status() == mediaway::Status::NoDevice;
}

bool writeFile(const std::string& path, const mediaway::Bytes& bytes) {
  std::ofstream out(path, std::ios::binary);
  if (!out) return false;
  out.write(reinterpret_cast<const char*>(bytes.data()),
            static_cast<std::streamsize>(bytes.size()));
  return static_cast<bool>(out);
}

}  // namespace

int main() {
  try {
    // ---- GPU device + Screen: display 0, 30 fps ------------------------------
    mediaway::device::GpuDevice gpu = mediaway::device::GpuDevice::create(
        {std::nullopt, /*videoSupport=*/true, /*debugLayer=*/false});

    mediaway::device::ScreenCapture screen;
    try {
      screen = mediaway::device::ScreenCapture::open({0, {1, 30}, gpu.handle()});
    } catch (const mediaway::Error& error) {
      if (!isExpectedUnavailable(error)) throw;
      std::cout << "screen capture unavailable: " << error.what()
                << " - exiting.\n";
      return EXIT_SUCCESS;
    }
    const mediaway::device::CaptureInfo& info = screen.info();
    std::cout << "screen geometry: " << info.width << 'x' << info.height
              << " @ " << info.frameRate.num << '/' << info.frameRate.den
              << '\n';

    // ---- Microphone: 48 kHz; drain-only (not wired into this MP4's tracks) ---
    std::optional<mediaway::device::AudioCapture> mic;
    try {
      mic.emplace(mediaway::device::AudioCapture::open({0, 48000}));
    } catch (const mediaway::Error& error) {
      if (!isExpectedUnavailable(error)) throw;
      std::cerr << "microphone unavailable: " << error.what()
                << " - screen only\n";
    }

    // ---- Encode at the negotiated geometry, sharing the same GPU device -----
    std::optional<mediaway::encoder::AutoVideoEncoder> encoder;
    try {
      encoder.emplace(mediaway::encoder::AutoVideoEncoder::open({
          mediaway::Codec::H264,
          info.width, info.height,
          info.frameRate,
          mediaway::PixelFormat::Bgra8,  // Screen delivers BGRA8, not NV12
          gpu.handle(),
      }));
    } catch (const mediaway::Error& error) {
      if (!isExpectedUnavailable(error)) throw;
      std::cout << "no GPU-input encoder available for " << info.width << 'x'
                << info.height << ": " << error.what() << " - exiting.\n";
      return EXIT_SUCCESS;
    }
    mediaway::encoder::EncodeSession session = std::move(*encoder).begin();

    // ---- Record ~3 s, bridging screen frames straight into the encoder ------
    const auto deadline =
        std::chrono::steady_clock::now() + std::chrono::seconds(3);
    std::size_t framesWritten = 0;
    std::size_t micFrames = 0;
    try {
      while (std::chrono::steady_clock::now() < deadline) {
        if (session.writeFrameFromDesktopCapture(screen)) {
          ++framesWritten;
        } else {
          std::this_thread::sleep_for(std::chrono::milliseconds(4));
        }
        if (mic) {
          while (std::optional<mediaway::device::AudioFrame> audio =
                     mic->pollFrame()) {
            ++micFrames;
          }
        }
      }
    } catch (const mediaway::Error& error) {
      if (!isExpectedUnavailable(error)) throw;
      // Same known dev-machine limitation as camera/encode_to_mp4's own
      // encoder — the WMF/DX11 backend accepts a GpuDevice-configured open
      // but rejects the GPU-input encode path once frames actually start
      // flowing (adr/pipeline/0002-gpu-frame-input-c-abi.md).
      std::cout << "GPU-input encode unsupported on this backend ("
                << error.what() << ") - exiting.\n";
      return EXIT_SUCCESS;
    }
    if (mic) mic->close();
    screen.close();

    if (framesWritten == 0) {
      std::cout << "screen capture opened but delivered no frames within the "
                   "deadline - exiting.\n";
      return EXIT_SUCCESS;
    }

    mediaway::Bytes mp4 = std::move(session).finish();
    std::cout << "bridged " << framesWritten << " real screen frame(s) (+ "
              << micFrames << " mic frames drained, not muxed) -> "
              << mp4.size() << " bytes\n";
    if (!writeFile("screen_out.mp4", mp4)) {
      std::cerr << "failed to write screen_out.mp4\n";
      return EXIT_FAILURE;
    }
    std::cout << "wrote screen_out.mp4\n";
    return EXIT_SUCCESS;
  } catch (const mediaway::Error& error) {
    std::cerr << "mediaway error: " << error.what() << " (status "
              << static_cast<int>(error.status()) << ")\n";
    return EXIT_FAILURE;
  }
}

File: bindings/python/examples/pipeline/screen_record.py
"""Screen + mic -> H.264 encode -> MP4 (out_screen.mp4).

Real Zero-Copy Screen capture + the capture-to-encode bridge: a `GpuDevice`
(adr/0007-gpu-device-factory.md) drives real Screen capture, and
`EncodeSession.write_frame_from_desktop_capture` (adr/pipeline/0005) polls +
pushes each frame in one native call — no intermediate `VideoFrame`, Zero-Copy
for Screen's GPU-backed frames. `AutoVideoEncoder.pick(..., pixel_format=BGRA8,
gpu_device=gpu)` negotiates the GPU-input-capable path before the bridge is
ever called (DXGI delivers BGRA8, not the NV12 default).

On a machine whose encoder backend accepts a GPU-configured open but rejects
the GPU-input path once frames actually start flowing (a real, pre-existing
WMF/DX11 limitation — see the Rust `gpu_write_frame_smoke.rs` test and the
C/Node.js/C# `screen_record` siblings), this gracefully skips instead of
crashing.

The audio side is not wired into the output MP4 as a second track yet — mic
PCM is drained (proves capture works), not muxed, same as `ScreenRecord.cs`'s
C# equivalent. `camera_record.py`'s two-track remux is a separate flow this
file doesn't duplicate.
"""

from contextlib import nullcontext
from pathlib import Path
import time

from mediaway import (
    AudioCapture,
    AutoVideoEncoder,
    CaptureUnsupportedError,
    DeviceUnavailableError,
    EncodeSession,
    GpuDevice,
    MediawayError,
    PixelFormat,
    Rational,
    VideoCapture,
)

RECORD_SECONDS = 3.0


def fmt_rational(r: Rational) -> str:
    """Render a Rational as a compact "num/den" string."""
    return f"{r.num}/{r.den}"


def main() -> None:
    try:
        gpu = GpuDevice.create(video_support=True)
    except (CaptureUnsupportedError, DeviceUnavailableError) as err:
        print(f"GPU device unavailable: {err} — exiting.")
        return

    with gpu:
        try:
            screen = VideoCapture.open("screen", index=0, frame_rate=Rational(1, 30), gpu_device=gpu)
        except (CaptureUnsupportedError, DeviceUnavailableError) as err:
            print(f"screen capture unavailable: {err} — exiting.")
            return

        with screen:
            width, height = screen.size()
            fps = screen.frame_rate()
            print(f"screen geometry: {width}x{height} @ {fmt_rational(fps)}")

            try:
                encoder = AutoVideoEncoder.pick(
                    width=width,
                    height=height,
                    frame_rate=fps,
                    pixel_format=PixelFormat.BGRA8,
                    bitrate_bps=8_000_000,
                    gpu_device=gpu,
                )
            except MediawayError as err:
                print(f"no GPU-input encoder available for {width}x{height}: {err} — exiting.")
                return
            print(f"picked encoder: {encoder.name} ({encoder.codec.name})")

            mic = None
            try:
                mic = AudioCapture.open("mic", 48000)
                print(f"mic open: {mic.sample_rate()} Hz, {mic.channels()} ch")
            except DeviceUnavailableError as err:
                print(f"mic unavailable ({err}); recording video only")

            with mic or nullcontext(), EncodeSession(encoder) as session:
                deadline = time.monotonic() + RECORD_SECONDS
                frames_written = 0
                pcm_bytes = 0
                try:
                    while time.monotonic() < deadline:
                        if session.write_frame_from_desktop_capture(screen):
                            frames_written += 1
                        else:
                            time.sleep(0.004)

                        if mic is not None:
                            pcm = mic.poll_pcm(timeout=0.0)
                            if pcm is not None:
                                pcm_bytes += len(pcm)
                except MediawayError as err:
                    print(f"GPU-input encode unsupported on this backend ({err}) — exiting.")
                    return

                if frames_written == 0:
                    print("screen capture opened but delivered no frames within the deadline — exiting.")
                    return

                mp4 = session.finish()  # terminal: consumes the session

    out_path = Path("out_screen.mp4")
    out_path.write_bytes(mp4)
    print(
        f"bridged {frames_written} real screen frame(s) over {RECORD_SECONDS:.0f} s "
        f"(drained {pcm_bytes} PCM bytes) -> {out_path} ({len(mp4)} bytes)"
    )


if __name__ == "__main__":
    main()
